tool/tsh/resolve_default_addr.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type raceResult struct {
+	addr string
+	err  error
+}
+
+// logResponseBody reads and dumps a response body to the log at the supplied
+// level. Note that it is still the caller's responsibility to close the body
+// stream.
+func logResponseBody(level logrus.Level, bodyStream io.Reader) {
+	if log.Logger.Level < level {
+		return
+	}
+
+	// NB: `ReadAll()` will time out (or be cancelled) according to the
+	//     context originally supplied to the request that initiated this
+	//     response, so no need to have an independent reading timeout
+	//     here.
+	body, err := ioutil.ReadAll(bodyStream)
+	if err != nil {
+		// This is only for debugging purposes, so it's safe to just give up here.
+		log.WithError(err).Debug("Failed to read body stream")
+		return
+	}
+
+	log.Logf(level, "Response body: %q", body)
+}
+
+// raceRequest drives an HTTP request to completion and posts the results back
+// to the supplied channel.
+func raceRequest(ctx context.Context, cli *http.Client, addr string, waitgroup *sync.WaitGroup, results chan<- raceResult) {
+	defer waitgroup.Done()
+
+	target := fmt.Sprintf("https://%s/webapi/ping", addr)
```

Comment:
What is the solution to that? I could spawn separate http & https racers I suppose.

Reply:
... although the webclient is just going to plug whtever we find straight into an `https` URL anyway. See:

https://github.com/gravitational/teleport/blob/2b7acc846e9142162d549ffbe66073c84abd5595/api/client/webclient/webclient.go#L70

Reply:
Hmm this is a good point. I'm only superficially familiar with `--insecure-no-tls` flag but I wonder if it exists solely for scenarios where proxy runs behind a TLS-terminating load balancer so the actual public proxy endpoint clients connect to is still TLS (i.e. proxy in "insecure no TLS" mode is not supposed to be connected to directly). If that's the case, then it's ok to just use https. @webvictim or @russjones maybe, do you folks know?

Reply:
Your reasoning is correct - `--insecure-no-tls` disables the TLS part of Teleport's web listener and just makes it speak plain HTTP, but is only intended to be used for backend -> Teleport connections. `tsh` and Teleport agents joining the cluster will still be expecting to communicate over HTTPS - it's just that the HTTPS will be terminated in front of Teleport rather than inside it.

While we're on the topic, though, it's worth checking how the `--insecure` flag to `tsh` and Teleport agents works as that skips all TLS verification, but I'm unsure whether it also allows the use of an HTTP connection rather than HTTPS when combined with `--insecure-no-tls` on the web listener.

Reply:
As far as `tsh` is concerned, it blithely assumes `https` everywhere that I can see. There are at least half a dozen places in the `webclient` that have constants that start "https://".